vertx-consul-client/src/main/java/io/vertx/ext/consul/Watch.java

```diff
@@ -0,0 +1,133 @@
+   * This maps to the <a href="https://www.consul.io/docs/agent/http/kv.html#single">/v1/kv/&lt;key&gt;</a> API internally.
+   *
+   * @param key the key
+   * @param vertx the {@code Vertx} instance
+   * @param options the options to create underlying Consul client
+   * @return the {@code Watch} instance
+   */
+  static Watch<KeyValue> key(String key, Vertx vertx, ConsulClientOptions options) {
+    return new WatchImpl.Key(key, vertx, options);
+  }
+
+  /**
+   * Creates {@code Watch} to monitoring the list of available services.
+   * The underlying Consul client will be created with default options.
+   * This maps to the <a href="https://www.consul.io/docs/agent/http/catalog.html#catalog_services">/v1/catalog/services</a> API internally.
+   *
+   * @param vertx the {@code Vertx} instance
+   * @return the {@code Watch} instance
+   */
+  static Watch<ServiceList> services(Vertx vertx) {
+    return services(vertx, new ConsulClientOptions());
+  }
+
+  /**
+   * Creates {@code Watch} to monitoring the list of available services.
+   * This maps to the <a href="https://www.consul.io/docs/agent/http/catalog.html#catalog_services">/v1/catalog/services</a> API internally.
+   *
+   * @param vertx the {@code Vertx} instance
+   * @param options the options to create underlying Consul client
+   * @return the {@code Watch} instance
+   */
+  static Watch<ServiceList> services(Vertx vertx, ConsulClientOptions options) {
+    return new WatchImpl.Services(vertx, options);
+  }
+
+  /**
+   * Creates {@code Watch} to monitoring the nodes providing the service.
+   * The underlying Consul client will be created with default options.
+   * This maps to the <a href="https://www.consul.io/docs/agent/http/health.html#health_service">/v1/health/service/&lt;service&gt;</a> API internally.
+   *
+   * @param service the service name
+   * @param vertx the {@code Vertx} instance
+   * @return the {@code Watch} instance
+   */
+  static Watch<ServiceEntryList> service(String service, Vertx vertx) {
+    return service(service, vertx, new ConsulClientOptions());
+  }
+
+  /**
+   * Creates {@code Watch} to monitoring the nodes providing the service.
+   * This maps to the <a href="https://www.consul.io/docs/agent/http/health.html#health_service">/v1/health/service/&lt;service&gt;</a> API internally.
+   *
+   * @param service the service name
+   * @param vertx the {@code Vertx} instance
+   * @param options the options to create underlying Consul client
+   * @return the {@code Watch} instance
+   */
+  static Watch<ServiceEntryList> service(String service, Vertx vertx, ConsulClientOptions options) {
+    return new WatchImpl.Service(service, vertx, options);
+  }
+
+  /**
+   * Set the result handler. As data is changed, the handler will be called with the result.
+   *
+   * @param handler the result handler
+   * @return reference to this, for fluency
+   */
+  @Fluent
+  Watch<T> setHandler(Handler<AsyncResult<T>> handler);
```

Comment:
> is the handler call once ? or for each upcoming values ?

for each upcoming values

> What kind of failure will be reported here ?

* "connection refused" if consul is unavailable
* "not found" if watched keys not exists
* etc